### PR TITLE
Effect UI untangling

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: *(direct link to the issue)*

Many steps toward extraction of the ui-independent parts of destructive effect
processing into a library.

One minor change in observable behavior:  certain modal dialogs may open while
an effect dialog is open (from sidebar or from the main menus).  These may now be
positioned differently, no longer centered as before above the effect dialog.  Preserving
the old behavior would add unnecessary bother to the implementation.

Also removes some, but not all, remaining state still carried in Effect objects that we
have called "stateless" (enough for realtime processing purposes) and so makes
more functions const (but adds some more const_casts in stateful effects).

Also improves terminology.  EffectUIValidator is renamed EffectEditor and
EffectUIClientInterface becomes EffectUIServices.

EffectUIServices is an interface of const member functions only, with one exception,
and they take const EffectPlugin & as context.  Just as a stateless Effect is a factory
for Instances, so now EffectUIServices is a factory for Editors.

Concrete effect classes inherit both EffectUIServices and EffectPlugin but EffectPlugin
now has no UI related functions in it.  The EffectUIEditor interface of an effect is
discovered where needed by a dynamic_cast.  Passing the effect again as the EffectPlugin
argument allows EffectUIServices not to be based on EffectPlugin, and enables some
common implementations of its methods to query the effect via EffectPlugin for some
other needed information.

For QA:  test these things
- Preview play of any effect in the destructive effect dialog -- including Amplify and Sliding stretch
- Showing and hiding of non-modal effect dialog; exist of program while it still shows
- Modal dialogs for effect options (from the manage menu) and for import/export -- as noted, positioning may be different, but let's not care about that change.  (Ladspa, AU, VST, VST3, LV2, and at least one built-in)
- Thoroughly testing dialogs of Echo; the Amplify effect; and Normalize
- 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/EffectBase.h [hunk 1]
```diff
@@ -37,10 +38,17 @@ class AUDACITY_DLL_API EffectBase /* not final */
 
    void SetTracks(TrackList *pTracks) { mTracks = pTracks; }
 
+   //! Called when Preview() starts, to allow temporary effect state changes
+   /*!
+    default returns a null
+    @return will undo its effects in its destructor before Preview() finishes
+    */
+   virtual std::any BeginPreview(const EffectSettings &settings);
```
Comment (crsib) [inline src/effects/EffectBase.h:44]: If it is called only by Preview, why is it public? It feels that it should be protected instead.

Comment (crsib) [reply src/effects/EffectBase.h:44]: Actually, this is a rare case where I don't disagree with `any` :-)

File: src/effects/Effect.cpp [hunk 2]
```diff
@@ -155,18 +155,17 @@ int Effect::ShowClientInterface(wxWindow &parent, wxDialog &dialog,
    EffectUIValidator *, bool forceModal)
 {
    // Remember the dialog with a weak pointer, but don't control its lifetime
```
Comment (crsib) [inline src/effects/Effect.cpp]: Comment appears to be out of date

File: src/effects/EffectBase.cpp [hunk 3]
```diff
@@ -390,10 +391,6 @@ void EffectBase::Preview(EffectSettingsAccess &access, bool dryOnly)
             std::dynamic_pointer_cast<EffectInstanceEx>(MakeInstance())
          )
             pInstance->Init();
-
-      // In case any dialog control depends on mT1 or mDuration:
-      if ( mUIDialog )
```
Comment (crsib) [inline src/effects/EffectBase.cpp:388]: I wonder why this branch is missing now

File: src/effects/EffectUI.cpp [hunk 4]
```diff
@@ -755,8 +755,12 @@ void EffectUIHost::OnPlay(wxCommandEvent & WXUNUSED(evt))
       if (!TransferDataFromWindow())
          return;
       
-      mEffectUIHost.Preview(*mpAccess, false);
-      
+      auto updater = [this]{ TransferDataToWindow(); };
+      mEffectUIHost.Preview(*mpAccess, updater, false);
+      // After restoration of settings and effect state:
+      // In case any dialog control depends on mT1 or mDuration:
+      updater();
```
Comment (crsib) [inline src/effects/EffectUI.cpp:771]: Oh, nvm question about the branch

File: src/effects/Effect.cpp [hunk 5]
```diff
@@ -68,11 +68,6 @@ Effect::Effect()
 
 Effect::~Effect()
 {
-   // Destroying what is usually the unique Effect object of its subclass,
-   // which lasts until the end of the session.
-   // Maybe there is a non-modal realtime dialog still open.
-   if (mHostUIDialog)
```
Comment (crsib) [inline src/effects/Effect.cpp:74]: I wonder if should revisit the modality of destructive dialogs though. It appears that we have broken a common workflow by making destructive application dialogs modal. 

This won't really matter when automation is introduced, so probably it doesn't really matters.

File: src/EffectPlugin.cpp [hunk 6]
```diff
@@ -25,9 +25,9 @@ bool EffectInstanceEx::Init()
 }
 
 EffectUIValidator::EffectUIValidator(
-   EffectUIClientInterface &effect, EffectSettingsAccess &access)
-   : mEffect{effect}
-   , mAccess{access}
+   EffectUIServices &services, EffectSettingsAccess &access)
+   : mServices{ services }
```
Comment (crsib) [inline src/EffectPlugin.cpp]: Minor suggestion - let's call it mUIServices maybe?

File: src/effects/EffectBase.h (hunk 1 above)
Comment (Paul-Licameli) [reply src/effects/EffectBase.h:44]: Protected makes sense.

std::any is just type erasure for things that can copy and may have nontrivial destructors.

My type erased pipeline class also uses it.

File: src/effects/Effect.cpp (hunk 2 above)
Comment (Paul-Licameli) [reply src/effects/Effect.cpp]: I think I removed it at some late commit in this long sequence?

File: src/effects/EffectBase.cpp (hunk 3 above)
Comment (Paul-Licameli) [reply src/effects/EffectBase.cpp:388]: See the finally

File: src/EffectPlugin.cpp (hunk 6 above)
Comment (Paul-Licameli) [reply src/EffectPlugin.cpp]: Ok

**Review by crsib — APPROVED** [review]
A good step towards toolkit neutrality for the effect.

**Comment by Paul-Licameli** [comment]
Rebased and force pushed with no changes or conflicts

**Comment by Paul-Licameli** [comment]
Rebased onto recent master, with conflicts resolved in CMakeLists.txt files and in include directives

**Comment by Paul-Licameli** [comment]
Rebased without change onto more recent master.

**Comment by dozzzzer** [comment]
@Paul-Licameli 
Moving back to To Do: loading a VST effect causes deadlock in both real-time and destructive application

STR:
1. Have some audio
2. Select a VST effect from the main Effect menu OR
3. Select a VST effect in the sidebar and try to open its dialog

Act. result: Audacity hangs completely

**Comment by Paul-Licameli** [comment]
@crsib please see how the new commits first turn the stack overflow into a build failure, then fix VST effects.

Once these are understood I want to squash them earlier in the branch.

**Review by crsib — APPROVED** [review]

**Comment by dozzzzer** [comment]
The issue above bisects to https://github.com/audacity/audacity/commit/95b7997ebb56e914d185cc3ed07e2dc55b7d5699
@Paul-Licameli

**Comment by Paul-Licameli** [comment]
> The issue above bisects to [95b7997](https://github.com/audacity/audacity/commit/95b7997ebb56e914d185cc3ed07e2dc55b7d5699)
> @Paul-Licameli

If so, then this branch is not the cause of it, but an issue should be opened about this recent regression and fixed as soon as I can.

**Comment by Paul-Licameli** [comment]
Rebased onto recent master to resolve small conflicts, but no other changes

**Comment by dozzzzer** [comment]
Verified

**Pull request merged**